2020/4/passport/src/main.rs

```diff
@@ -0,0 +1,248 @@
+                }
+
+                &_ => {}
+            }
+        }
+        Passport {
+            birth_year: birth_year,
+            issue_year: issue_year,
+            expiration_year: expiration_year,
+            height: height,
+            hair_color: hair_color,
+            eye_color: eye_color,
+            passport_id: passport_id,
+            country_id: country_id,
+        }
+    }
+
+    pub fn is_valid(&self) -> bool {
+        match &self.birth_year {
+            Some(byr) => is_valid_year(&byr, 1920, 2002) && match &self.issue_year {
+                Some(iyr) => is_valid_year(&iyr, 2010, 2020) && match &self.expiration_year {
+                    Some(eyr) => is_valid_year(&eyr, 2020, 2030) && match &self.height {
+                        Some(h) => is_valid_height(h) && match &self.hair_color {
+                            Some(hcl) => is_valid_hair_color(hcl) && match &self.eye_color {
+                                Some(ecl) => is_valid_eye_color(ecl) && match &self.passport_id {
+                                    Some(pid) => is_valid_passport_id(pid),
+                                    None => false,
+                                },
+                                None => false,
+                            },
+                            None => false,
+                        },
+                        None => false,
+                    },
+                    None => false,
+                },
+                None => false,
+            },
+            None => false,
+        }
+    }
+}
+
+fn is_valid_year(value: &str, min: usize, max: usize) -> bool {
+    if value.len() != 4 {
+        return false;
+    }
+
+    let v = value.parse::<usize>().unwrap();
+    min <= v && v <= max
+}
+
+fn is_valid_hair_color(value: &str) -> bool {
+    if value.len() != 7 {
+        return false;
+    }
+
+    let mut chars = value.chars();
+    if chars.next().unwrap() != '#' {
+        return false;
+    }
+
+    for _ in 1..6 {
+        if !chars.next().unwrap().is_digit(16) {
+            return false;
+        }
+    }
+    true
+}
+
+fn is_valid_eye_color(value: &str) -> bool {
+    match value {
+        "amb" => true,
+        "blu" => true,
+        "brn" => true,
+        "gry" => true,
+        "grn" => true,
+        "hzl" => true,
+        "oth" => true,
+        _ => false,
+    }
+}
+
+fn is_valid_passport_id(value: &str) -> bool {
+    value.len() == 9 && value.chars().all(|c| c.is_digit(10))
+}
+
+fn is_valid_height(value: &str) -> bool {
+    if value.ends_with(&"cm") {
+        return is_between(value, 150, 193);
+    } else if value.ends_with(&"in") {
+        return is_between(value, 59, 76);
+    }
+
+    false
+}
+
+fn is_between(value: &str, min: usize, max: usize) -> bool {
+    let len = value.len();
+    let v = value[0..len - 2].parse::<usize>();
+    match v {
+        Ok(vv) => vv >= min && vv <= max,
+        Err(_) => false,
+    }
+}
+
+#[test]
```

Comment:
A pattern I use for "grouping" tests in Rust (generally not necessary unless you have a number of helper functions, which you don't here) is to make a sub-module that only compiles in test:

```rust
#[cfg(test)]
mod test {
    // Put all the parent module names in scope
    use super::*;

    #[test]
    pub fn my_test() { ... }

    pub fn my_helper() { ... }
}
```

The only trick there is that you're in a new module so you need things like `use super::*` to bring the parent module (the rest of the file) in to scope, and if you have `use` blocks at the top and need to `use` the same things in your test, you have to do it again. A sub-module effectively creates a new "file" as far as the compiler is concerned.

You *do* still get the advantage of being able to see private values from the parent module in your tests though (which you don't get if you move the tests to a completely separate file). In Rust, visibility is all based on the module hierarchy. Any module can see `pub` values from any other module, but **only** *child modules* can see non-`pub` values. The equivalent of "internals visible to" in Rust is to put your tests *in* the module-under-test **or** put them in a *child module* of the module-under-test.

Reply:
Nice, will consider a module from now on. I can't remember how I got into the separate file last year.

Reply:
A separate file for testing is definitely used in Rust, but generally for integration tests that test multiple components using public apis.